File: win64deploy.pl
#!/usr/bin/perl -w

use strict;
use Cwd;
use File::Copy;
use File::Path;

######################################################################
# EXTERNAL PATHS
my $qt_path = "c:/Qt";
my $qbin_path = $qt_path . "/5.12.5/mingw73_64/bin";
my $qssl_path = $qt_path . "/Tools/QtCreator/bin";
#my $openssl_path = "c:/OpenSSL-Win32";
my $msvc_path = "c:/Program Files (x86)/Microsoft Visual Studio 14.0/VC";

######################################################################
# INTERNAL PATHS
my $cschem_buildpath = "build-cschem-Desktop_Qt_5_12_5_MinGW_64_bit-Release/release";
my $cpcb_buildpath = "build-cpcb-Desktop_Qt_5_12_5_MinGW_64_bit-Release/release";
my $release_path = "release-cschem-x64";

######################################################################

my @envpath = split(/:/, $ENV{PATH});
my @usepath;
for (@envpath) {
	push @usepath, $_ unless /Anaconda/;
}
$ENV{PATH} = join(":", @usepath);	

$msvc_path =~ s/\//\\/g;
$ENV{VCINSTALLDIR} = $msvc_path;

die "cschem executable not found" unless -f "$cschem_buildpath/cschem.exe";
die "cpcb executable not found" unless -f "$cpcb_buildpath/cpcb.exe";


File::Path::remove_tree($release_path) if -d $release_path;
File::Path::make_path($release_path);

system("$qbin_path/windeployqt --dir $release_path "
       . " --compiler-runtime $cschem_buildpath/cschem.exe")
  and die "Failed to get cschem deployment";

system("$qbin_path/windeployqt --dir $release_path "
       . " --compiler-runtime $cpcb_buildpath/cpcb.exe")
  and die "Failed to get cpcb deployment";

File::Copy::copy("$cschem_buildpath/cschem.exe", "$release_path/");
File::Copy::copy("$cpcb_buildpath/cpcb.exe", "$release_path/");

sub sslcopy {
  my $pth = shift;
  my $fn = shift;
  print "$fn not found in $pth - no https support\n" unless -f "$pth/$fn";
  File::Copy::copy("$pth/$fn", "$release_path/"); 
}

#sslcopy($qssl_path, "ssleay32.dll", "$release_path/");
#sslcopy($qssl_path, "libeay32.dll", "$release_path/");
#sslcopy($openssl_path, "libcrypto-1_1.dll", "$release_path/");
#sslcopy($openssl_path, "libssl-1_1.dll", "$release_path/");

print "Now run 'cschem-x64.iss' using Inno Setup.\n";

File: tools/win64deploy.py
#!/usr/bin/python3

import os
import shutil
import glob

#%% EXTERNAL PATHS
qt_path = "c:/Qt/5.12.10/msvc2017_64"
qbin_path = qt_path + "/bin"
msvc_path = r"C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC"

#%% INTERNAL PATHS
root = "C:/Users/Wagenaar/Documents/Progs/cschem"
cschem_build = root + "/build-cschem-Desktop_Qt_5_12_10_MSVC2017_64bit-Release/release"
cpcb_build = root + "/build-cpcb-Desktop_Qt_5_12_10_MSVC2017_64bit-Release/release"
release_path = root + "/release-x64"
bin_path = release_path + '/bin'

#%%
envpath = os.environ['PATH'].split(";")
usepath = [p for p in envpath if "anaconda" not in p.lower()]
os.environ['PATH'] = ';'.join(usepath)
os.environ['VCINSTALLDIR'] = msvc_path

#%%
if not os.path.exists(cschem_build + "/cschem.exe"):
    raise Exception("cschem executable not found")
if not os.path.exists(cpcb_build + "/cpcb.exe"):
    raise Exception("cpcb executable not found")

#%%
if os.path.exists(release_path):
    shutil.rmtree(release_path)
os.mkdir(release_path)

def copydir(name):
    files = glob.glob(f"{root}/{name}/*")
    dst = release_path + f"/{name}"
    os.mkdir(dst)
    for f in files:
        f.replace('\\', '/')
        if os.path.isdir(f):
            idx = f.index(name)
            copydir(f[idx:])
        else:
            shutil.copy(f, dst)

os.system(f"{qbin_path}/windeployqt --release --dir {bin_path} {cpcb_build}/cpcb.exe")
os.system(f"{qbin_path}/windeployqt --release --dir {bin_path} {cschem_build}/cschem.exe")
shutil.copy(f"{cpcb_build}/cpcb.exe", bin_path)
shutil.copy(f"{cschem_build}/cschem.exe", bin_path)
copydir("symbols")
copydir("pcb-outlines")

print("Now run 'tools/cschem-x64.iss' using Inno Setup.")
